File: python/백준/이분탐색/2467 용액.py
N = int(input())
nums = list(map(int, input().split()))
nums.sort()

min_num = [0, 0, int(1e9)]  # [값1, 값2, 현재 최솟값(절댓값)]

def search(left, right):
    while left < right:
        s = nums[left] + nums[right]

        # 최솟값 갱신 (절댓값 기준)
        if abs(s) < abs(min_num[2]):
            min_num[0], min_num[1], min_num[2] = nums[left], nums[right], abs(s)

        if s == 0:
            print(nums[left], nums[right])
            return True  # 정확히 0이면 바로 종료

        if s > 0:
            right -= 1
        else:
            left += 1
    return False

# 이 문제는 한 번만 돌리면 됨
found = search(0, N-1)
if not found:
    print(min_num[0], min_num[1])

File: cpp/백준/이분탐색/2467 용액.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    vector<long long> nums(N);
    for (int i = 0; i < N; ++i) cin >> nums[i];
    sort(nums.begin(), nums.end());

    long long best1 = nums[0], best2 = nums[1];
    long long bestAbs = llabs(nums[0] + nums[1]);

    int left = 0, right = N - 1;
    bool found = false;
    while (left < right) {
        long long s = nums[left] + nums[right];
        if (llabs(s) < bestAbs) {
            best1 = nums[left];
            best2 = nums[right];
            bestAbs = llabs(s);
        }
        if (s == 0) {
            cout << nums[left] << " " << nums[right] << "\n";
            found = true;
            break;
        }
        if (s > 0) --right;
        else ++left;
    }
    if (!found) {
        cout << best1 << " " << best2 << "\n";
    }
    return 0;
}
